### ISSUE TITLE
GC per thread malloc + parallel mark

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#1548]
I have disabled the per thread malloc in the GC as it's causing problems. We want to bring it back though, so let's find out what the problem is and how we can bring per thread malloc back. Is the GC really the problem?

First, for the errors, running this test script:
```sh
#!/usr/bin/env sh
ACTONC=../dist/bin/actonc

${ACTONC} --dev rts_db/ddb_test_server.act
${ACTONC} --dev rts_db/ddb_test_client.act
${ACTONC} --dev rts_db/test_tcp_client.act
FAIL=0
for I in $(seq 200); do
    killall actondb ddb_test_server ddb_test_client test_tcp_client >/dev/null 2>&1 || true
    echo "-- TEST $I FAIL: $FAIL"
    rts_db/test_tcp_client --verbose || FAIL=$((FAIL+1))
done

echo "FAIL: $FAIL"
```

so we get 200 test runs
- with per thread malloc & parallel mark we get about 30% failures.
- with only parallel mark, we get 0 failures or maybe 1 (not percent, just 1 failure)
- with only per thread malloc (no parallel mark), we get about 2% failures

It seems per thread malloc and parallel mark makes it worse when combined.

**Comment by plajjan** [comment]
And as usual with GC related problems it might not be the GC at all, just that it helps unearth some other problem. After disabling the GC, the CI errors went away on Linux (or low enough that it's not very noticeable anymore) while for MacOS I still saw errors. Picked up my MacBook and started looking.

I suspect a bug related to DNS lookups. We already have stdlib_auto/test_net_dns.act that does DNS lookups and that seems to work just fine, but continuously running test/rts_db/test_tcp_client results in a lot of SIGILL on MacOS. We are doing DNS lookups both in the program under test, ddb_test_client, but also in the test_tcp_client app test orchestrator itself, which seem to account for most of the failures and since it's the app test runner, its crashing have weird effects and cannot be easily handled. I don't know why but it somehow seems to be hanging, so running it in a bash for loop, I still need to ctrl+c it and in CI we hit the timeout in GitHub Actions (6 hours) rather than it just failing.
    
I've looked quickly at the backtrace and it seems we crash in _lookup_a__on_resolve. The DNS query has failed with status -3008 and error message "DNS lookup error: unknown node or service". Now obviously a failed DNS lookup shouldn't lead to a crash, but it seems like the on_error callbak that we store in the data struct passed via the libuv DNS query struct is NULL. I don't see how can that be, since it is unconditionally set when we send in the request. Perhaps a bug in libuv? Maybe add some asserts here? Annoying that we have to guard against bugs in these APIs, I'd expect them to be fairly mature by now.

Anyway, after avoiding DNS lookups in our test suite (8aa5026bc679d209a382b0dbb33472205dcac5e2) the problem I saw with per thread malloc are completely gone! So, good news, we can put per thread malloc back in... and the bad news is that DNS resolution is broken.

**Comment by plajjan** [comment]
I noticed problems on Linux related to uv_fs_scandir(). The fix was to tell the GC to IGNORE_FRRE and that's what we currently run with. I suspect that there is a memory allocation done in libc which isn't redirected to GC_MALLOC and so when we later to free on it we get GC_FREE which barfs because there is no GC header on the allocation.

The free is called in libuv and is explicitly coded to do a regular free(). libuv has pluggable mallocs but for this allocation it explicitly reached for the regular free() which I interpret as libuv developers expect this allocation to be done with regular malloc but since we do runtime redirection we manage to redirect the libuv free but not the (presumably) libc malloc, leading to the problem.

When we remove link time redirection of malloc & friends, I think this bug will fix itself.

I can't help but wonder if the DNS errors we saw are of a related kind, so the removal of link time redirection for malloc will fix it. That should then make it possible to re-enable per thread malloc...

**Comment by plajjan** [comment]
I realize we do have per thread malloc. It's been enabled again ever since 1f3159169ec2a7d68880b43f9c02dda76925291c when I switched to the upstream build of bdwgc which has a new build.zig (that we wrote & upstreamed). Previously we had used a build.zig in our fork of the bdwgc repo that was tailored for our Acton use, so we had per thread alloc disabled. Switching to the upstream version, we got the defaults from the upstream which has per thread alloc enabled as well as parallel marking enabled.

The test.sh programs runs well in this combination, so this is effectively fixed since quite some time now :P  closing
